### ISSUE TITLE
Sky renders in front of volumetric fog and clouds

### ISSUE THREAD

**Issue by TokisanGames** [TokisanGames/Sky3D#12]
The sky renders in front of volumetric fog.

Volumetric fog & fog volume w/ Sky3D screen space fog disabled.

![{1EDF0512-3974-49C3-9DF5-E3C37AA0F096}](https://github.com/user-attachments/assets/bdec2142-4c7e-4a27-bba2-d0267c234197)

Volumetrics w/ Sky3D fog enabled has artifacts around buildings. The screenspace fog also flickers at certain angles when both are enabled. 

![{61D72847-5750-4288-9379-CD3D7B87588C}](https://github.com/user-attachments/assets/20a7c8d8-a173-49f9-967d-8ce28ed77f2d)

**Comment by TokisanGames** [comment]
### How it currently works

The sky renders on a large sphere mesh instance, which I believe is stuck to the projection matrix (camera) via the vertex shader.

The fog renders on a quad mesh, which is stuck to the screen via the vertex shader.
 
### Possible paths forward

A non-comprehensive list.

1. Identify rendering priority of the volumetric fog and figure out how to get the screen space shaders in an appropriate order. I haven't had success playing with the priorities.

2. Port the shaders into the new godot compositor. I don't know anything about this.

3. Port the shaders into the godot sky material. We're already running a sky material for reflective light, so even though we have clouds in the sky, reflections show a clear blue sky. It would be nice to have clouds and sky provide accurate reflective light intensity and color (currently it's a reasonable approximation), and show in reflections.
 
4. Disconnect the sky from the projection matrix and place it in the world by disabling skip_vertex_transform and vertex() setting position, and setting the sphere mesh size very large. This requires the sphere be larger than your world, and be connected to the camera, and camera far distance be larger than the sphere. This shows promise. Volumetrics work with or without the screen space sky3d fog shader

Volumetrics on, Sky3D fog off.

![{D404BD5A-6E28-407C-952D-BD50EF451208}](https://github.com/user-attachments/assets/2275b214-6a51-42f2-be07-e6db73f80efb)

Volumetric fog density 0, fog volume density 1, colored blue, Sky3D fog on

![{7FC5CA39-4FAC-4D5D-ABD2-6D93C7DE5B88}](https://github.com/user-attachments/assets/3ba7a845-89d6-48a9-bed1-1a479bf798cc)

**Comment by sourceonfire** [comment]
Add `DEPTH = -1.0`; and removing `skip_vertex_transform` seems to fix the problem with the skydome

Another thing, from v1.0.1 to v1.3 (where I removed the repo) there was a change in the shader transformation. I don't remember why but I know it worked better.

```
vec4 vert = vec4(VERTEX, 0.0);
POSITION =  PROJECTION_MATRIX * VIEW_MATRIX* MODEL_MATRIX * vert;
POSITION.z = POSITION.w;
```

**Comment by sourceonfire** [comment]
- If you want custom fog you will have to use the composer, fog is always a problem with transparent objects.
- Currently the reason to keep the skydome is in this problem: https://github.com/godotengine/godot/issues/81871 but we lose the reflections of the PBR, so we can send a small texture of the sky to the world environment and update it every x time.
- Godot's fog gets the color of the sky automatically with skyshaders, when sending the texture you can forget about the custom fog, unless you need it.

**Comment by sourceonfire** [comment]
This seems to fix the problem with the skydome and volumetric fog for me.

```
vec4 vert = vec4(VERTEX.xyz, 0.0) ;
vert.xyz *= 1000.0;
VERTEX = (MODELVIEW_MATRIX * vert).xyz; // This is required to apply fog if you use skip_vertex_transform
POSITION = PROJECTION_MATRIX * MODELVIEW_MATRIX * vert;
POSITION.z = 0.0;
```

**Comment by TokisanGames** [comment]
@sourceonfire Hi J, thanks for the suggestion. This does address the sky, but it makes fog volumes a weird. The sky sometimes shows through the volume at half capacity, sometimes not, depending on the camera angle. So far the best and most stable option I've tried is option 4 above.

**Comment by sourceonfire** [comment]
> [@sourceonfire](https://github.com/sourceonfire) Hi J, thanks for the suggestion. This does address the sky, but it makes fog volumes a weird. The sky sometimes shows through the volume at half capacity, sometimes not, depending on the camera angle. So far the best and most stable option I've tried is option 4 above.


Yes, the error occurs when the camera is outside the boundaries of the dome, so it has to be very large.

**Comment by TokisanGames** [comment]
Fixed by #36
